pkg/pipeline/common_steps.go

```diff
@@ -69,29 +74,49 @@ func addTenantLabel(obj PipelineObject, data *ExecutionContext) ExecutionResult
 
 func updateObject(obj PipelineObject, data *ExecutionContext) ExecutionResult {
 
-	oldResourceVersion := obj.GetObjectMeta().GetResourceVersion()
-
 	rtObj, ok := obj.(runtime.Object)
 	if !ok {
-		return ExecutionResult{
-			Abort: true,
-			Err:   fmt.Errorf("object ist not a valid runtime.object: %v", obj.GetObjectMeta().GetName()),
+		return ExecutionResult{Err: errors.New("object is not a runtime object")}
+	}
+
+	if !specAndMetaEqual(obj, data.originalObject) {
+		err := data.Client.Update(context.TODO(), rtObj)
+		if err != nil {
+			if k8serrors.IsConflict(err) {
+				return ExecutionResult{Abort: true}
+			}
+			return ExecutionResult{Err: err}
 		}
 	}
 
-	updatedObj := rtObj.DeepCopyObject()
-	err := data.Client.Update(context.TODO(), updatedObj)
-	if err != nil {
-		return ExecutionResult{Err: err}
+	if !equality.Semantic.DeepEqual(data.originalObject.GetStatus(), obj.GetStatus()) {
+
+		origRtObj := data.originalObject.(runtime.Object)
+
+		err := data.Client.Status().Patch(context.TODO(), rtObj, client.MergeFrom(origRtObj))
```

Comment:
What's the reason for using `Patch()` here?

Reply:
See above explanation about the merging.

Reply:
But that's not the status we updated, but the one we deep copied at the beginning of the reconcile?

Reply:
You're right, it gets merged from the wrong instance.

Reply:
Explains why it needs two reconciles currently to be applied correctly.

Reply:
I just found a weird behaviour. If I rewrite the snipped to look like this:

```
deepCopy := rtObj.DeepCopyObject()

if !specAndMetaEqual(obj, data.originalObject) {
		err := data.Client.Update(context.TODO(), rtObj)
		if err != nil {
			if k8serrors.IsConflict(err) {
				return ExecutionResult{Abort: true}
			}
			return ExecutionResult{Err: err}
		}
	}

	if !equality.Semantic.DeepEqual(data.originalObject.GetStatus(), obj.GetStatus()) {

		origRtObj := data.originalObject.(runtime.Object)

		err := data.Client.Status().Patch(context.TODO(), rtObj, client.MergeFrom(origRtObj))
		if err != nil {
			return ExecutionResult{Err: err}
		}
	}
```

The object WON'T contain the status on the k8s API. Even though stepping through with the debugger shows that `rtObj` and `deepCopy` both contain the status before the `Status().Patch()` call.

If using the snipped as it's currently committed it works as expected. There is something weird going on with that copy.

Reply:
This only makes sense if `data.Client.Update(context.TODO(), rtObj)` doesn't update the status of the `rtObj` object.

Reply:
Just tested it, `data.Client.Update(context.TODO(), rtObj)` will update everything in `rtObj` to the same state as applied to the cluster except the `Status` field.

But it does not explain why `data.Client.Status().Patch(context.TODO(), rtObj, client.MergeFrom(deepCopy))` won't set the actual status though. As I can verify it being called with the correct values.

Reply:
Going through the `Patch()` code it looks like they only get apply IF there's a difference between both given objects. This is not the case in my example with deepCopy. So I propose to leave it as it is now with the `origObj`. This way we don't have to merge the fields ourselves and can leverage the `Patch()` method further.

Reference for the called merge code: https://github.com/evanphx/json-patch/blob/v4.5.0/merge.go#L223